pkg/core/tekton/builder/pipeline.go

```diff
@@ -0,0 +1,131 @@
+package builder
+
+import (
+	"github.com/tektoncd/pipeline/pkg/apis/pipeline/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// PipelineBuilder holds the tekton pipeline definition.
+type PipelineBuilder struct {
+	Pipeline v1beta1.Pipeline
+}
+
+// NewPipelineBuilder creates a Pipeline with default values.
+func NewPipelineBuilder(name, namespace string) *PipelineBuilder {
+	b := &PipelineBuilder{}
+	b.Pipeline = v1beta1.Pipeline{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: namespace,
+		},
+	}
+	return b
+}
+
+// PipelineMeta sets the Meta structs of the Pipeline.
+// Any number of MetaOp modifiers can be passed.
+func (b *PipelineBuilder) PipelineMeta(ops ...MetaOp) {
+	for _, op := range ops {
+		switch o := op.(type) {
+		case ObjectMetaOp:
+			o(&b.Pipeline.ObjectMeta)
+		case TypeMetaOp:
+			o(&b.Pipeline.TypeMeta)
+		}
+	}
+}
+
+// PipelineDescription sets the Description on the Pipeline spec.
+func (b *PipelineBuilder) PipelineDescription(description string) {
+	b.Pipeline.Spec.Description = description
+}
+
+// PipelineParam adds a ParamSpec to the Pipeline spec.
+func (b *PipelineBuilder) PipelineParam(name string, description string, defaultValue ...string) {
+	param := v1beta1.ParamSpec{
+		Name:        name,
+		Description: description,
+	}
+	if len(defaultValue) > 0 {
```

Comment:
Why is `defaultValue` an array when you always use just first item? What about `*string`, and passing nil for no default value?

Reply:
I think this is a personal choice, there are different ways to implement optional parameters to functions in Go, this is just another way that I prefer than passing `nil`.

Reply:
I understand you might like checking for length instead of checking for nil (or passing empty list instead of nil), the weird part is that it's either empty list or list with one item. Really, what sense does it make to use a list argument when you do not expect it can have more than one items?

Reply:
Golang doesn't support optional function parameter or default values for function parameters, and there are a few ways to work around this problem. Using a _variadic function_ like you do here is one of them, but it has several disadvantages over the other methods:
* it only works well with one optional parameter. It gets confusing with 2 or more optional parameters
* it's not forwards compatible: if you need to add more optional parameters in the future, you need to change the function signature in a way that breaks existing calls
* it's already a bit confusing, because the function accepts several values, but you ignore everything but the first one, the name is `defaultValue`, but it's actually an array and so on

You should really avoid this approach and use one of the other workarounds:
* different functions with different parameters: a `PipelineParamWithDefaultValue` function would look great here
* use _functional options_ for optional parameters, same as you did earlier with `PipelineMeta`. This may be a bit overkill for a single optional parameter, but otherwise it's the best approach I've seen for optional parameters.

There is a lot of info on the subject of optional parameters with golang online. This is one that I found insightful and partially applicable to this case: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis

Reply:
After reviewing some more code changes, I have to say that I'm surprised that so far you've only encountered the case where you only have _one_ optional parameter. How is that possible ?

Reply:
Thanks the provided link is very informational.

I have added additional functions for those cases.

Reply:
Thanks. BTW, your reluctance to pass `nil` as a parameter value is also shared by the author: _I believe that we, as Go programmers, should work hard to ensure that nil is never a parameter that needs to be passed to any public function._